### PR TITLE
PM can request modifications to fixed-price invoice

### LINKED ISSUE #2453: PM can request modifications to fixed-price invoice
## Additional context
This is part of epic #2352.

## Describe the solution you'd like in Hypha
- PM opens the invoice
- reviews it
- PM decides the invoice needs to be modified
- PM updates the invoice status and sends a message to the contractor requesting changes to the invoice
- the invoice is then put on hold until the contractor resubmits the invoice

## Notifications
When the PM updates the invoice status to "request changes":
- the contractor receives a notification requesting they update the invoice

## Priority
- High (keeping you from completing day-to-day tasks)

## Affected roles 
- Staff
- Applicants
- Finance

## Ideal deadline
June 2021

### Demo
https://user-images.githubusercontent.com/1899506/125316942-e4b9fd00-e338-11eb-99de-c78a308d7116.mp4

### Wireframes

**6.1**

The PM logs in and sees the table of invoices to be reviewed. This table shows:
- the date of the status and the status 
- the invoice reference - made up of the project name and the reference number generated by Hypha, and
- the total value amount

The PM chooses to review “invoice maya summit 58”.

**6.2**

Invoice reference Maya Summit (58) is opened. The PM reviews the invoice. 

The contractor has invoiced for “Designing a website” (Deliverable 1) and “Hosting a website” (Deliverable 2). They have invoiced $6,000 for Deliverable 1 when it should it $5,000. (The PM knows this from the drop-down of deliverables, but also from the contract)

This is incorrect and so the PM cannot approve the invoice. To fix this they will update the invoice status to “request changes” and the invoice will be sent back to the contractor with a message.

**6.3**

This then opens the modal "Update invoice status" window.

**6.4**

From here they choose “request changes”.
(The finance related approvals are unavailable to then as they do not have those permissions.)

The other options available are “Approved by PM” or “Declined”.

**6.5**

Then the PM writes a message explaining what is incorrect with the invoice, and how the contractor needs to fix it. This message written here will later be shown in the Activity feed.

When finished writing the message, they click update status.

**6.6**

The invoice status is then updated to Requested changes, with the date the status change happened.

If the PM wants to see the message they’ve written to the contractor, they can navigate to the project, and click on activity feed and see the message there.

**6.7**

From the all projects list the PM can navigate to the Maya Summit project page.

**6.8**

Once the PM has opened the project page, they can click on the activity feed.

**6.9**

In the activity feed, the PM can see that the status has been updated, and they can also see the message which has been sent to the contractor.

### PR DESCRIPTION
**Pull request by sks444**
Fixes https://github.com/HyphaApp/hypha/issues/2453

### CODE REVIEW

**Comment by sks444** [comment]
@frjo, this looks good to me.

Major check list for this are:

- Create both type of activity comment(communication tab)and action(activity feed.).  
- Send email to applicant about invoice status update.

Demo: https://github.com/HyphaApp/hypha/issues/2453#issuecomment-971261893

**Comment by sks444** [comment]
Will be handling the contractor editing and resubmitting https://github.com/HyphaApp/hypha/issues/2619 in a separate pr.

File: hypha/apply/activity/messaging.py [hunk 1]
```diff
@@ -372,6 +372,10 @@ def handle_skipped_report(self, report, **kwargs):
         else:
             return "Marked a Report as required"
 
+    def handle_update_invoice_status(self, invoice, **kwargs):
+        invoice_status_change = _(f'Updated Invoice status to: {invoice.status_display}.')
```
Comment (frjo) [inline hypha/apply/activity/messaging.py]: f-strings are really nice until you need to translate the strings. Please use the older format-function instead.

`_('Updated Invoice status to: {status}.').format(status=invoice.status_display)`

Here we also try to make the variable in the string easy to understand for the translator. The Translator will only se "Updated Invoice status to: {status}." and from that understand how to best translate.

File: hypha/apply/activity/templates/messages/email/invoice_status_updated.html [hunk 2]
```diff
@@ -2,15 +2,18 @@
 
 {% load i18n %}
 {% block content %}
-{% blocktrans with title=source.title date_from=invoice.date_from date_to=invoice.date_to %}An {{ ORG_SHORT_NAME }} staff member has updated your invoice for {{ title }} for period {{ date_from }} to {{ date_to }}.{% endblocktrans %}
+
+{% blocktrans with title=source.title date_from=invoice.date_from date_to=invoice.date_to %}An {{ ORG_SHORT_NAME }} staff member has updated your invoice for {{ date_from }} for period to {{ date_to }}.{% endblocktrans %}
```
Comment (frjo) [inline hypha/apply/activity/templates/messages/email/invoice_status_updated.html]: I do not think this change is intended.

**Review by frjo — CHANGES_REQUESTED** [review]
After a change has been requested and the applicant resaves the invoice the status need to change. I think we for now can go with `SUBMITTED` for now. We might need to add a specific `RESUBMITTED` state, a bit unsure about that. 

Some minor code comments as well.

**Comment by frjo** [comment]
@chaitu210 Nice work on this one!

Moving forward can you please use `git rebase origin/main` on feature branches instead of using merge. This give us a cleaner history and give us fewer merge conflicts overall.

**Comment by sks444** [comment]
>After a change has been requested and the applicant resaves the invoice the status need to change. I think we for now can go with SUBMITTED for now. We might need to add a specific RESUBMITTED state, a bit unsure about that.

@frjo Yeah, we will be handling contractor editing the invoice and status change in a separate pr https://github.com/HyphaApp/hypha/issues/2619; that's work in progress.

File: hypha/apply/activity/messaging.py (hunk 1 above)
Comment (chaitu210) [reply hypha/apply/activity/messaging.py]: Updated changes as mentioned in the comment.

File: hypha/apply/activity/templates/messages/email/invoice_status_updated.html (hunk 2 above)
Comment (chaitu210) [reply hypha/apply/activity/templates/messages/email/invoice_status_updated.html]: Updated changes as mentioned in the comment.

File: hypha/apply/projects/views/payment.py [hunk 3]
```diff
@@ -53,6 +56,21 @@ def get_form_kwargs(self):
 
     def form_valid(self, form):
         response = super().form_valid(form)
+        if form.cleaned_data['comment']:
+            invoice_status_change = _(f'<p>Invoice status updated to : {self.object.status_display}. </p>')
```
Comment (frjo) [inline hypha/apply/projects/views/payment.py]: F-strings do not work with translations. Please change to use format().

File: hypha/apply/projects/views/payment.py [hunk 4]
```diff
@@ -53,6 +56,21 @@ def get_form_kwargs(self):
...
+            comment = '<p>{}</p>'.format(form.cleaned_data['comment'])
```
Comment (frjo) [inline hypha/apply/projects/views/payment.py]: Here I would prefer an F-string.

**Review by frjo — CHANGES_REQUESTED** [review]

**Review by frjo — APPROVED** [review]

**Pull request merged**